### PR TITLE
Check whether a cluster is using JBOD when populate_disk_info is true

### LINKED ISSUE #1353: Error message should state that "populate_disk_info" should be set to true only if JBOD is used in Kafka cluster
Users can query Kafka cluster load along with `populate_disk_info`. For example:

`GET http://<Cruise Control host>:<port>/kafkacruisecontrol/load?populate_disk_info=true`

If the Kafka cluster is not using JBOD. It returns:

`com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException: java.lang.IllegalStateException: Missing disk information for disk...`

The error message should be more specific and make it clear that `"populate_disk_info" should be set to true only if JBOD is used in Kafka cluster`

### PR DESCRIPTION
**Pull request by Lincong**
This PR resolves #1353 

Notes for reviewers:

The implementation is based on the assumption that if a `Broker` has multiple `Disk`s, it must be using JBOD. Please correct me if this assumption is incorrect. Thanks!

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java [hunk 1]
```diff
@@ -202,6 +202,10 @@ public boolean isAlive() {
     return _state != State.DEAD;
   }
 
+  public boolean isUsingJBOD() {
+    return _diskByLogdir.size() > 1;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java:206]: 1. Can we add a JavaDoc?
2. Nit: `_diskByLogdir.size() > 1` -> `!_diskByLogdir.isEmpty()`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java [hunk 2]
```diff
@@ -64,13 +66,29 @@ protected BrokerStats getResult() throws Exception {
         return cachedBrokerStats;
       }
     }
+    if (_populateDiskInfo && !isClusterUsingJBOD()) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:69]: Can we avoid checking `_populateDiskInfo` here again -- i.e. can't we do the following?
```
} else if (!isClusterUsingJBOD()) {
   ...
}
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java [hunk 3]
```diff
@@ -64,13 +66,29 @@ protected BrokerStats getResult() throws Exception {
...
+    // If and only if all brokers in the cluster are using JBOD, the cluster is considered to be using JBOD
+    for (Broker broker : clusterModel.brokers()) {
+      if (!broker.isUsingJBOD()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:91]: Nit: extra space.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java [hunk 4]
```diff
@@ -64,13 +66,29 @@ protected BrokerStats getResult() throws Exception {
         return cachedBrokerStats;
       }
     }
+    if (_populateDiskInfo && !isClusterUsingJBOD()) {
+      throw new UserRequestException("populate_disk_info = true when Kafka cluster is NOT using JBOD");
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:70]: Can we refrain from (1) using hardcoded parameter names and (2) emphasis via capitalization -- i.e. maybe we can use something like:
```
throw new UserRequestException(String.format("Cannot set %s=true for non-JBOD Kafka clusters.", POPULATE_DISK_INFO_PARAM));
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java [hunk 5]
```diff
@@ -64,13 +66,29 @@ protected BrokerStats getResult() throws Exception {
...
 
+  private boolean isClusterUsingJBOD() throws Exception {
+    ClusterModel clusterModel = _kafkaCruiseControl.loadMonitor().clusterCapacity();
+    // If and only if all brokers in the cluster are using JBOD, the cluster is considered to be using JBOD
+    for (Broker broker : clusterModel.brokers()) {
+      if (!broker.isUsingJBOD()) {
+        return false;
+      }
+    }
+    return true;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:88]: Nit: This seems like a good candidate to replace with stream API for conciseness:
```
return clusterModel.brokers().stream().allMatch(Broker::isUsingJBOD);
```

**Review by efeg — COMMENTED** [review]
@Lincong Thanks for the patch!
Left some comments.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java (hunk 2 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:69]: Good idea!

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java (hunk 3 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:91]: Fixed

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java (hunk 4 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:70]: Good idea

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java (hunk 1 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java:206]: If we use `!_diskByLogdir.isEmpty()`, does that mean if `_diskByLogdir` is not empty, the broker must be using JBOD?

I thought `_diskByLogdir.size() == 1` means that the broker is using non-JBOD and `_diskByLogdir .size() > 1`  means that the broker is using JBOD.

Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java]: > If we use `!_diskByLogdir.isEmpty()`, does that mean that the broker is using JBOD if and only if `_diskByLogdir` is NOT empty?

Correct -- please see how we `_diskByLogdir` is initialized in constructor.

Note that if a broker has 1 disk (i.e. non-JBOD), then the cluster model does not add another layer of abstraction -- i.e. keeps track of replicas per broker as opposed to replicas per disk in the broker, which would always be the same disk/logDir in a non-JBOD broker.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java [hunk 6]
```diff
@@ -7,9 +7,12 @@
 import com.linkedin.kafka.cruisecontrol.KafkaCruiseControl;
 import com.linkedin.kafka.cruisecontrol.async.progress.OperationProgress;
 import com.linkedin.kafka.cruisecontrol.exception.KafkaCruiseControlException;
+import com.linkedin.kafka.cruisecontrol.model.Broker;
 import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
 import com.linkedin.kafka.cruisecontrol.monitor.ModelCompletenessRequirements;
+import com.linkedin.kafka.cruisecontrol.servlet.UserRequestException;
 import com.linkedin.kafka.cruisecontrol.servlet.parameters.ClusterLoadParameters;
+import com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java:15]: Nit: Maybe we can avoid having the boilerplate `ParameterUtils.XXX` if we static import just the `POPULATE_DISK_INFO_PARAM`.
```
import static com.linkedin.kafka.cruisecontrol.servlet.parameters.ParameterUtils.POPULATE_DISK_INFO_PARAM;
```

**Review by efeg — COMMENTED** [review]
Thanks for the update -- added just one more comment.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java (hunk 6 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/handler/async/runnable/LoadRunnable.java]: OK. Fixed

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java (hunk 1 above)
Comment (Lincong) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java]: Got it. Thanks

**Review by efeg — APPROVED** [review]
Lgtm!

**Pull request merged**